### PR TITLE
[Demo] Get gym data from server Pull #666 instead

### PR DESCRIPTION
**Pull request by JohanJu**
## Description
Printing data about Pokemon(name, cp, moves) and players(name, level) in gyms and gym name.
This is more of a demo on what is possible to get from gyms, use #666 when that is fully implemented as it follows the code convention.

## Motivation and Context
#605

## How Has This Been Tested?
Ran for 10 min 4 thread without crash

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [x ] Breaking change (fix or feature that would cause existing functionality to change)
Increase api calls

## Checklist:

- [? ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by chuyskywalker** [comment]
This is a really good idea, but we "repetively" scan gyms waaaay too much to do that right now. However, I'm thinking about some changes where we'd setup dedicated gym scanners and we could, potentially, integrate this to do a slow burn population of gym data.

**Comment by karololszak** [comment]
I did a similar thing with `api.fort_details` (to get pokestops / gyms name, description and images). 
To prevent rescanning same forts, I implemented a simple persistent cache (in file) and I just cached based on fort id and location. But I think the best way to do it would be to check against our db (use it as a cache). It wouldn't work in this example however, because players in gyms change - fort_details don't.
To speed it up a bit, you can lower the sleep to 0.3s (when doing the scan with 1 acc).

**Comment by JohanJu** [comment]
I think that we may put a timeout in the "cache" like 2min. I am unsure about the sleep time as we don't want to get banned, do we have data on how often we can scan without banning?

I have no front end experience, can someone say which file/class/functions are used transporting data from the database to the popup in browser. 

Also if I would put it in the database, is it just a data column och do we want new tables and fancy joins to be able to filter on players etc.

**Comment by Studio-XP** [comment]
`function loadRawData ()` in map.js:1357 loads the data from the address raw_data via ajax.  The info in raw_data is pushed via app.py

**Comment by JohanJu** [comment]
Thanks Studio-XP

Output with timeout 30 sec. (without Pokemon print)
```
2016-08-14 14:23:15,878 [ search_worker_2][        search][    INFO] Search step 4 beginning (queue size is 3)
2016-08-14 14:23:18,035 [ search_worker_3][        models][    INFO] Upserted 0 pokemon, 107 pokestops, and 14 gyms
2016-08-14 14:23:18,037 [ search_worker_3][        search][    INFO] Search step 5 beginning (queue size is 2)
Gym: QlikView Obelisk
Gym: Environmental Engineering W
Gym: Golden Arch
Gym: Victoriastadion
Gym: Rymdskeppet
Gym: Fyrklöverns Trädgård
Gym: Ideon Gateway
Gym: Web
2016-08-14 14:23:28,747 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 94 pokestops, and 15 gyms
2016-08-14 14:23:28,747 [ search_worker_0][        search][    INFO] Search step 6 beginning (queue size is 1)
Gym: Kapitalet
2016-08-14 14:23:33,471 [ search_worker_1][        models][    INFO] Upserted 0 pokemon, 103 pokestops, and 15 gyms
2016-08-14 14:23:33,472 [ search_worker_1][        search][    INFO] Search step 7 beginning (queue size is 0)
2016-08-14 14:23:35,782 [ search_worker_2][        models][    INFO] Upserted 1 pokemon, 110 pokestops, and 13 gyms
2016-08-14 14:23:35,783 [ search_worker_2][        search][    INFO] Search step 1 beginning (queue size is 6)
Gym: Forrest Siluette
2016-08-14 14:23:37,398 [ search_worker_3][        models][    INFO] Upserted 1 pokemon, 112 pokestops, and 12 gyms
2016-08-14 14:23:37,400 [ search_worker_3][        search][    INFO] Search step 2 beginning (queue size is 5)
2016-08-14 14:23:41,443 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 114 pokestops, and 15 gyms
2016-08-14 14:23:41,443 [ search_worker_0][        search][    INFO] Search step 3 beginning (queue size is 4)
2016-08-14 14:23:45,265 [ search_worker_1][        models][    INFO] Upserted 0 pokemon, 103 pokestops, and 14 gyms
2016-08-14 14:23:45,266 [ search_worker_1][        search][    INFO] Search step 4 beginning (queue size is 3)
2016-08-14 14:23:48,233 [ search_worker_2][        models][    INFO] Upserted 0 pokemon, 107 pokestops, and 14 gyms
2016-08-14 14:23:48,235 [ search_worker_2][        search][    INFO] Search step 5 beginning (queue size is 2)
2016-08-14 14:23:52,690 [ search_worker_3][        models][    INFO] Upserted 0 pokemon, 94 pokestops, and 15 gyms
2016-08-14 14:23:52,691 [ search_worker_3][        search][    INFO] Search step 6 beginning (queue size is 1)
Gym: QlikView Obelisk
Gym: Environmental Engineering W
Gym: Golden Arch
Gym: Victoriastadion
Gym: Rymdskeppet
Gym: Fyrklöverns Trädgård
Gym: Ideon Gateway
Gym: Web
2016-08-14 14:24:03,446 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 103 pokestops, and 15 gyms
2016-08-14 14:24:03,447 [ search_worker_0][        search][    INFO] Search step 7 beginning (queue size is 0)
Gym: Kapitalet
Gym: Afrodite
2016-08-14 14:24:06,543 [ search_worker_1][        models][    INFO] Upserted 1 pokemon, 110 pokestops, and 13 gyms
2016-08-14 14:24:06,544 [ search_worker_1][        search][    INFO] Search step 1 beginning (queue size is 6)
Gym: Forrest Siluette
2016-08-14 14:24:08,085 [ search_worker_2][        models][    INFO] Upserted 1 pokemon, 112 pokestops, and 12 gyms
2016-08-14 14:24:08,086 [ search_worker_2][        search][    INFO] Search step 2 beginning (queue size is 5)
```

**Comment by JohanJu** [comment]
Alt text

**Comment by Studio-XP** [comment]
Pulling now to test

**Comment by prusswan** [comment]
If you are implementing fetching fort details, can you add a link to the pokestop popup to fetch their details? Since pokestops details don't involve distance check.

**Comment by JohanJu** [comment]
@prusswan I am not implementing fort(pokestop) details, what do you want from the fort, it's name?

**Comment by Zerotask** [comment]
I get the following errors

> [2016-08-14 17:53:20,290] ERROR in app: Exception on /raw_data [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1639, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1625, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/bin/PokemonGo-Map/pogom/app.py", line 96, in raw_data
    d['gyms'] = Gym.get_gyms(swLat, swLng, neLat, neLng)
  File "/bin/PokemonGo-Map/pogom/models.py", line 281, in get_gyms
    for g in query:
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2967, in __iter__
    return iter(self.execute())
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2960, in execute
    self._qr = ResultWrapper(model_class, self._execute(), query_meta)
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2656, in _execute
    return self.database.execute_sql(sql, params, self.require_commit)
  File "/usr/local/lib/python2.7/dist-packages/playhouse/shortcuts.py", line 208, in execute_sql
    cursor.execute(sql, params or ())
  File "/usr/lib/python2.7/dist-packages/MySQLdb/cursors.py", line 226, in execute
    self.errorhandler(self, exc, value)
  File "/usr/lib/python2.7/dist-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
OperationalError: (1054, "Unknown column 't1.info' in 'field list'")
2016-08-14 17:53:20,290 [        Thread-7][           app][   ERROR] Exception on /raw_data [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1639, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1625, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/bin/PokemonGo-Map/pogom/app.py", line 96, in raw_data
    d['gyms'] = Gym.get_gyms(swLat, swLng, neLat, neLng)
  File "/bin/PokemonGo-Map/pogom/models.py", line 281, in get_gyms
    for g in query:
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2967, in __iter__
    return iter(self.execute())
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2960, in execute
    self._qr = ResultWrapper(model_class, self._execute(), query_meta)
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2656, in _execute
    return self.database.execute_sql(sql, params, self.require_commit)
  File "/usr/local/lib/python2.7/dist-packages/playhouse/shortcuts.py", line 208, in execute_sql
    cursor.execute(sql, params or ())
  File "/usr/lib/python2.7/dist-packages/MySQLdb/cursors.py", line 226, in execute
    self.errorhandler(self, exc, value)
  File "/usr/lib/python2.7/dist-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
OperationalError: (1054, "Unknown column 't1.info' in 'field list'")

**Comment by JohanJu** [comment]
@Zerotask did you get it directly or after some time? I ran for 10min with 4 acc without problem.

**Comment by Zerotask** [comment]
I get it directly after start up. Isn't there a change missing in app.py?

`if request.args.get('gyms', 'true') == 'true':
            d['gyms'] = Gym.get_gyms(swLat, swLng, neLat, neLng)`

**Comment by JohanJu** [comment]
Why do app.py needs to be edited? Can you try with a clean clone form my repo https://github.com/JohanJu/PokemonGo-Map

**Comment by jnovack** [comment]
I disagree that timeout should be `now()+60`.  Just add `last_scanned` like gyms.  Don't add logic to the database layer.

**Comment by JohanJu** [comment]
@jnovack This will probably need some refactoring, I will look at last_scanned.  You are welcome to make a pull request as I'm just started with this project today and don't know your design.

**Comment by jnovack** [comment]
@karololszak Completely off topic, can you submit a PR to cache the gym/pokestop name/picture_url in the database?

@JohanJu if you want your PR to be accepted, you should stick close to the conventions already in place for the project.  Let the back-end take care of the logic (removing gyms via timeout) and let the database layer just keep the data, stale or not.

**Comment by Zerotask** [comment]
It's working when I add the column in the db manually. Great Job

**Comment by JohanJu** [comment]
@Zerotask thanks
@jnovack Will try, but it may take some time

File: pogom/models.py [hunk 1]
```diff
@@ -387,17 +389,36 @@ def parse_map(map_dict, step_location):
...
-                    'last_modified': datetime.utcfromtimestamp(
-                        f['last_modified_timestamp_ms'] / 1000.0),
-                }
+                gid = f['id']
+                info = ''
+                now = int(time.time())
+                if gid not in gyms_timeout or gyms_timeout[gid] < now:
+                    try:
+                        data = api.get_gym_details(gym_id=f['id'])['responses']['GET_GYM_DETAILS']
+                        info += data['name'] + '<br>'
```
Comment (jnovack) [inline pogom/models.py]: Do not add HTML code to the value returned.  Return a JSON object and let the browser/css/front-end deal with it.

**Comment by Zerotask** [comment]
Now I don't get any Pokemon nor Pokestops. Do you know what could be the reason for that?

> 2016-08-14 19:18:51,071 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:51,384 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:51,406 [ search_worker_0][        search][    INFO] Search step 7 beginning (queue size is 54)
> 2016-08-14 19:18:51,552 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:51,885 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:51,900 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:52,053 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:52,397 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:52,571 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:52,920 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:53,094 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:53,357 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:53,526 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:53,893 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:54,062 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:54,404 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:54,569 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:54,918 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:55,089 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:55,359 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:55,528 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:56,111 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:56,296 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:56,371 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:56,538 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:56,880 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)
> 2016-08-14 19:18:57,187 [ search_worker_0][        models][    INFO] Upserted 0 pokemon, 0 pokestops, and 0 gyms
> 2016-08-14 19:18:57,417 [ search_worker_0][        search][    INFO] Search step 8 beginning (queue size is 53)

**Comment by JohanJu** [comment]
@Zerotask ip/soft/hard ban maybe?

**Comment by Zerotask** [comment]
Login is working properly. How can I check a softban quickly? I currently use about 20 accounts / worker. Did you notice similar problems?

**Comment by JohanJu** [comment]
20 accounts? I would be more worried about ip ban. If you have a phone on the same ip you can try to use that and see if there are anything in the map. I'm going to sleep now, you can ask in the discord for more info about bans, it probably don't have much to do with this.

**Comment by Zerotask** [comment]
Login with my phone works, also checked 2 accounts with quickspin, it said that I have no softban.

**Comment by Studio-XP** [comment]
Not sure why, but occasionally the info data is blank. Is it because of how far away it is from the search zone?

blank gym info

**Comment by jnovack** [comment]
Gym visibility is 1km (25% of the time I can see 1.1km, 5% of the time I can see up to 1.2km, and others, but 99% within 1km).

However, Gym ability to get information is about 900m.  I consistently get issues close to 1km gyms.  

**For Gym Only Scans, this should be limited to 900m.**

**Comment by jnovack** [comment]
Here's an example search at .97km.  Notice the dirty Red Gym is over 1km away.

image

image

**Comment by prusswan** [comment]
I think @karololszak may have implemented what I had in mind. Originally I thought gym details is obtained from fort details as well.

**Comment by Zerotask** [comment]
Indeed I received an IP ban. The curious thing is that I got it after implementing your PR

**Comment by JohanJu** [comment]
@Zerotask This commit makes more api calls, So if you run 20 bot 24/7 you have to set some delays accordingly. (See comment 2 and 3)
In models.py you can search for time.sleep and set something.
The last commit (use last_modified) will hopefully lower the api calls but it will still be more than without this pull.

**Pull request closed without merging**